compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -819,7 +819,21 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         }
 
         match self.candidate_from_obligation(stack) {
-            Ok(Some(c)) => self.evaluate_candidate(stack, &c),
+            Ok(Some(c)) => {
+                let mut res = self.evaluate_candidate(stack, &c)?;
+                // If we erased any lifetimes, then we want to use
```

Comment:
Two things are holding me up on this PR:

1. Is this the right place for this, as opposed to (say) within `evaluate_candidate`?
2. Do we not have a test case?

Reply:
This is the only place that calls `evaluate_candidate`, so I think either places could work.

@ld has [done some work on minimizing this](https://github.com/rust-lang/rust/issues/83226#issuecomment-803064019), but AFAIK we don't yet have a standalone test for this.

Reply:
I've since then [reduced chalk](https://github.com/rust-lang/rust/issues/83115#issuecomment-803686837): it's shorter than that one, and without dependencies

Reply:
Can you move it to `evaluate_candidate` then? I can't imagine who would call `evaluate_candidate` and *not* want this change made. Also, @lqd's minimized issue seems small enough to include. r=me with those changes.